File: archive/c/c/zeckendorf.c
#include <errno.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

static void usage(void)
{
    fputs("Usage: please input a non-negative integer\n", stderr);
    exit(EXIT_FAILURE);
}

void zeckendorf(uint64_t n)
{
    if (n == 0)
    {
        putchar('\n');
        return;
    }

    uint64_t a = 1, b = 2;

    while (b <= n)
    {
        uint64_t next = a + b;
        a = b;
        b = next;
    }

    bool first = true;

    while (n > 0)
    {
        if (a <= n)
        {
            if (!first)
                fputs(", ", stdout);
            printf("%" PRIu64, a);
            n -= a;
            first = false;
        }

        uint64_t prev = b - a;
        b = a;
        a = prev;
    }

    putchar('\n');
}

int main(int argc, char *argv[])
{
    if (argc != 2)
        usage();

    const char *s = argv[1];
    char *end = NULL;

    errno = 0;
    uint64_t n = strtoull(s, &end, 10);

    if (*s == '\0' || *end != '\0' || errno == ERANGE || *s == '-')
        usage();

    zeckendorf(n);
    return EXIT_SUCCESS;
}

File: archive/r/ruby/zeckendorf.rb
# frozen_string_literal: true

USAGE = "Usage: please input a non-negative integer"

def usage!
  warn USAGE
  exit 1
end

def parse_input(str)
  usage! if str.nil? || str.strip.empty?

  n = Integer(str)
  usage! if n < 0

  n
rescue ArgumentError, NoMethodError
  usage!
end

def fibs_upto(n)
  fibs = [1, 2]
  fibs << fibs[-1] + fibs[-2] while fibs[-1] <= n
  fibs.pop if fibs[-1] > n
  fibs
end

def zeckendorf(n)
  return [] if n == 0

  fibs = fibs_upto(n)
  result = []

  i = fibs.length - 1

  while n > 0
    if fibs[i] <= n
      result << fibs[i]
      n -= fibs[i]
      i -= 2
    else
      i -= 1
    end
  end

  result
end

input = ARGV.first
n = parse_input(input)

puts zeckendorf(n).join(", ")
